shared/src/balancer/logexpmath.rs

```diff
@@ -0,0 +1,522 @@
+        4 => 800_000_000_000_000_000_000_i128,
+        5 => 400_000_000_000_000_000_000_i128,
+        6 => 200_000_000_000_000_000_000_i128,
+        7 => 100_000_000_000_000_000_000_i128,
+        8 => 50_000_000_000_000_000_000_i128,
+        9 => 25_000_000_000_000_000_000_i128,
+        10 => 12_500_000_000_000_000_000_i128,
+        11 => 6_250_000_000_000_000_000_i128,
+        _ => panic!("Constant not provided"),
+    }
+    .into()
+}
+fn constant_x_18(i: u32) -> I256 {
+    match i {
+        0 => 128_000_000_000_000_000_000_i128,
+        1 => 64_000_000_000_000_000_000_i128,
+        _ => panic!("Constant not provided"),
+    }
+    .into()
+}
+fn constant_a_20(i: u32) -> I256 {
+    match i {
+        2 => 7_896_296_018_268_069_516_100_000_000_000_000_i128,
+        3 => 888_611_052_050_787_263_676_000_000_i128,
+        4 => 298_095_798_704_172_827_474_000_i128,
+        5 => 5_459_815_003_314_423_907_810_i128,
+        6 => 738_905_609_893_065_022_723_i128,
+        7 => 271_828_182_845_904_523_536_i128,
+        8 => 164_872_127_070_012_814_685_i128,
+        9 => 128_402_541_668_774_148_407_i128,
+        10 => 113_314_845_306_682_631_683_i128,
+        11 => 106_449_445_891_785_942_956_i128,
+        _ => panic!("Constant not provided"),
+    }
+    .into()
+}
+fn constant_a_18(i: u32) -> I256 {
+    match i {
+        0 => {
+            I256::from_dec_str("38877084059945950922200000000000000000000000000000000000").unwrap()
+        }
+        1 => 6_235_149_080_811_616_882_910_000_000_i128.into(),
+        _ => panic!("Constant not provided"),
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Errors {
+    XOutOfBounds,
+    YOutOfBounds,
+    ProductOutOfBounds,
+    InvalidExponent,
+}
+
+pub fn pow(x: U256, y: U256) -> Result<U256, Errors> {
+    if y == U256::zero() {
+        return Ok(*UNSIGNED_ONE_18);
+    }
+    if x == U256::zero() {
+        return Ok(U256::zero());
+    }
+
+    let x_int256 = match I256::try_from(x) {
+        Ok(x) => x,
+        Err(_) => return Err(Errors::XOutOfBounds),
+    };
+
+    let y_int256 = if y < *MILD_EXPONENT_BOUND {
+        I256::try_from(y).unwrap()
+    } else {
+        return Err(Errors::YOutOfBounds);
+    };
+
+    let mut logx_times_y = if *LN_36_LOWER_BOUND < x_int256 && x_int256 < *LN_36_UPPER_BOUND {
+        let ln_36_x = _ln_36(x_int256);
+        (ln_36_x / *ONE_18) * y_int256 + ((ln_36_x % *ONE_18) * y_int256) / *ONE_18
+    } else {
+        _ln(x_int256) * y_int256
+    };
+    logx_times_y /= *ONE_18;
+
+    if !(*MIN_NATURAL_EXPONENT <= logx_times_y && logx_times_y <= *MAX_NATURAL_EXPONENT) {
+        return Err(Errors::ProductOutOfBounds);
+    }
+
+    exp(logx_times_y).map(|v| v.into_raw())
+}
+
+fn exp(mut x: I256) -> Result<I256, Errors> {
+    if !(x >= *MIN_NATURAL_EXPONENT && x <= *MAX_NATURAL_EXPONENT) {
+        return Err(Errors::InvalidExponent);
+    }
+
+    if x < I256::zero() {
+        return Ok((*ONE_18 * *ONE_18) / exp(-x)?);
```

Comment:
the *s in here looks awkwardly placed, spaced and separated...

I have also noticed a lot of `*` before variable and not between anything... is this something I should know about rust that I do not?

Reply:
My understanding of this is that `lazy_static` creates a new type for every `static ref` we define with it. This type can be dereferenced to the definition type, which is why I use the `*`.
I think it's like that because the thing defined by lazy_static is an object that takes care of initializing the actual object, and not the object itself.

Reply:
Oh right! I completely forgot about * for dereferencing. Silly me.